docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -1078,6 +1078,112 @@ public @interface CanWrite {
+import jakarta.ws.rs.HeaderParam;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.SecurityContext;
+import jakarta.ws.rs.core.UriInfo;
+
+public class SimpleBeanParam {
+
+    @QueryParam("query")
+    private String privateQuery;
+
+    @HeaderParam("Authorization")
+    String authorizationHeader;
+
+    @Context
+    SecurityContext securityContext;
+
+    @Context
+    UriInfo uriInfo;
+
+    String getPrivateQuery() {
+        return privateQuery;
+    }
+}
+----
+
+And let's say we have `SimplePermission` where we need to check `Authorization` header and principal name.
+That can be done like this:
+
+[source,java]
+----
+package org.acme.crud;
+
+import java.security.Permission;
+import java.util.Objects;
+
+public class SimplePermission extends Permission {
+
+    private final String authorizationHeader;
+    private final String principalName;
+
+    public SimpleBeanParamPermission(String permissionName, String authorizationHeader, String name, String privateQuery) { <1> <2>
+        super(permissionName);
+        this.authorizationHeader = authorizationHeader;
+        this.principalName = name;
+        if ("private-query".equals(privateQuery)) {
+            // your business logic
+        }
+    }
+
+    @Override
+    public boolean implies(Permission requiredPermission) {
+        if (requiredPermission instanceof SimpleBeanParamPermission that) {
+            return "MyAuthorization".equals(that.authorizationHeader)
+                && "admin".equals(that.principalName);
+        }
+        return false;
+    }
+
+    ...
+}
+----
+<1> Constructor parameters `authorizationHeader` and `principalName` corresponds to `SimpleBeanParam` fields with same names.
+<2> Constructor parameter `privateQuery` is accessed with a getter method.
+Quarkus always prefer method access, therefore, first we look for the `privateQuery` method, then for the `getPrivateQuery` method.
+If neither method can be found on the `SimpleBeanParam` class, Quarkus tries to find the `privateQuery` field.
+
+The `SimplePermission` can be used on your Jakarta REST resource like in the example below:
+
+[source,java]
+----
+package org.acme.crud;
+
+import io.quarkus.security.PermissionsAllowed;
+import jakarta.ws.rs.BeanParam;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+@Path("/hello")
+public class SimpleResource {
+
+    @PermissionsAllowed(value = "read", permission = SimplePermission.class,
+        params = { "param.authorizationHeader", "param.privateQuery", "param.securityContext.getUserPrincipal.name" }) <1>
+    @GET
+    public String sayHello(@BeanParam SimpleBeanParam param) {
+        return "Hello path is " + param.uriInfo.getPath();
+    }
+
+}
+----
+<1> `SecurityIdentity` principal is accessed from the `jakarta.ws.rs.core.SecurityContext`.
+Please note that the `name` part of the `param.securityContext.getUserPrincipal.name` expression matches the `SimplePermission` parameter `name`.
```

Comment:
People will eventually ask that we be able to say that `beanparam.name` should map to the `personName` permission constructor parameter…

Reply:
I agree, but I am happy to wait till someone ask :-).